source/opt/replace_invalid_opc.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "replace_invalid_opc.h"
+
+#include <bitset>
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status ReplaceInvalidOpcodePass::Process(ir::IRContext* c) {
+  InitializeProcessing(c);
+  bool modified = false;
+
+  if (context()->get_feature_mgr()->HasCapability(SpvCapabilityLinkage)) {
+    return Status::SuccessWithoutChange;
+  }
+
+  SpvExecutionModel execution_model = GetExecutionModel();
+  if (execution_model == SpvExecutionModelKernel ||
+      execution_model == SpvExecutionModelMax) {
```

Comment:
Is there a case when it should be equal SpvExecutionModelMax?

Reply:
This response applies to a few of the comments below.  The problem that can come up is if there are two entry points with different models, and there call trees are not disjoint.  A function in both call trees could be in a conflicting situation.  You need to keep the instruction because of one the first entry point, but it also needs to be removed for the second one.

This can lead to a mildly complicated analysis.  It can be done, but it complicates things.  Considering that our current set of shaders generates a single entry point per shader module, we did not consider it worth implementing.  So will just do a simple check (all entry points must have the same model).

For this check, we use SpvExecutionModelMax as a special value to indicate that the entry points have inconsistent model.  No shader can have the model SpvExecutionModelMax.  However, I will admit this is not future proof.

Reply:
I asking for a comment not because I don't understand what it is doing, but to improve readability. Consider how the code looks to somebody who sees it for the first time.

This looks much more readable.
SpvExecutionModel execution_model = GetExecutionModel();
if (execution_model == SpvExecutionModelKernel) {
  // Nothing to do.
  return;
}

if (execution_model == SpvExecutionModelMax) {
  // Multiple execution models in the module. Case not handled.
  return;
}

Reply:
Sorry, the original comment before out edited it seemed like you were asking under what circumstances `GetExecutionModel` would return ModelMax.  Now that I understand what you really meant, I think you have a good suggestion.  I'll fix that up.